api/tests/test_rate_limit.py

```diff
@@ -0,0 +1,246 @@
+from functools import partial
+
+import fakeredis.aioredis
+import pytest
+
+from src.rate_limit import (
+    RedisRateLimiter,
+    MAX_CONNECTIONS_PER_USER,
+    TooManyConnectionsException,
+    SERVER_LIVENESS_EXPIRATION_SECONDS,
+    RateLimiter,
+    MemoryRateLimiter,
+    MemoryRateLimiterStorage,
+    create_redis_rate_limiter,
+)
+
+
+@pytest.fixture
+async def redis(event_loop):
+    redis_instance = await fakeredis.aioredis.create_redis_pool()
+    yield redis_instance
+    redis_instance.close()
+    await redis_instance.wait_closed()
+
+
+@pytest.fixture
+def memory_rate_limiter_storage():
+    return MemoryRateLimiterStorage()
+
+
+@pytest.fixture
+def redis_rate_limiter_factory(redis):
+    return partial(create_redis_rate_limiter, redis=redis)
+
+
+@pytest.fixture
+def memory_rate_limiter_factory(memory_rate_limiter_storage):
+    async def fn(server_id: str):
+        return MemoryRateLimiter(server_id, memory_rate_limiter_storage)
+
+    return fn
+
+
+@pytest.fixture
+async def redis_rate_limiter(redis_rate_limiter_factory):
+    return await redis_rate_limiter_factory('server-id')
+
+
+@pytest.fixture
+async def memory_rate_limiter(memory_rate_limiter_factory):
+    return await memory_rate_limiter_factory('server-id')
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    'rate_limiter',
+    [
+        pytest.lazy_fixture('redis_rate_limiter'),
+        pytest.lazy_fixture('memory_rate_limiter'),
+    ],
+)
+async def test_acquire_connection(rate_limiter: RateLimiter):
```

Comment:
nit: the name of this function makes me think that you're testing that you can make a connection (which, I suppose it does), but the point of the test is that you _can't_ make a connection when you have the max amount of connections.